ext/LinearSolveEnzymeExt.jl

```diff
@@ -9,6 +9,47 @@ using Enzyme
 
 using EnzymeCore
 
+function EnzymeCore.EnzymeRules.forward(func::Const{typeof(LinearSolve.init)}, ::Type{RT}, prob::EnzymeCore.Annotation{LP}, alg::Const; kwargs...) where {RT, LP <: LinearSolve.LinearProblem}
+    @assert !(prob isa Const)
+    res = func.val(prob.val, alg.val; kwargs...)
+    if RT <: Const
+        return res
+    end
+    dres = func.val(prob.dval, alg.val; kwargs...)
+    dres.b .= res.b == dres.b ? zero(dres.b) : dres.b
+    dres.A .= res.A == dres.A ? zero(dres.A) : dres.A
+    if RT <: DuplicatedNoNeed
+        return dres
+    elseif RT <: Duplicated
+        return Duplicated(res, dres)
+    end
+end
+
+function EnzymeCore.EnzymeRules.forward(func::Const{typeof(LinearSolve.solve!)}, ::Type{RT}, linsolve::EnzymeCore.Annotation{LP}; kwargs...) where {RT, LP <: LinearSolve.LinearCache}
+    @assert !(linsolve isa Const)
+
+    A = deepcopy(linsolve.val.A) #mutates after function is applied
+    res = func.val(linsolve.val; kwargs...)
+    
+    if RT <: Const
+        return res
+    end
+    
+    dres = deepcopy(res)
+    invA = inv(A)
```

Comment:
never use `inv`

Reply:
addressed in d010911

Reply:
Not fully, see https://github.com/SciML/LinearSolve.jl/pull/416#discussion_r1382633907. Your version has an extra factorization and is using the wrong linear solver.

Reply:
addressed in 31745cc